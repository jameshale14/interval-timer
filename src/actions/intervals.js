import database from '../firebase/firebase'

//GET_INTERVALS
export const getIntervals = () => ({
  type: 'GET_INTERVALS',
})

//START_GET_INTERVALS

//CREATE_INTERVAL
export const createInterval = (interval) => ({
  type: 'CREATE_INTERVAL',
  interval
})

//START_CREATE_INTERVAL

//UPDATE_INTERVAL
export const updateInterval = (id, updates) => ({
  type: 'UPDATE_INTERVAL',
  id,
  updates
})

//START_UPDATE_INTERVAL

//DELETE_INTERVAL
export const deleteInterval = (id) => ({
  type: 'DELETE_INTERVAL',
  id
})

//START_DELETE_INTERVAL

export const startDeleteInterval = (id) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/intervals/${id}`).remove()
      .then(() => {
        dispatch(deleteInterval(id))
      })
  }
}

//SET_INTERVALS
export const setIntervals = (intervals) => ({
  type: 'SET_INTERVALS',
  intervals
})

//START_SET_INTERVALS
export const startSetIntervals = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/intervals`).once('value')
      .then((snapshot) => {
        const intervals = []
        snapshot.forEach((childSnapshot) => {
          intervals.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          })
        })
        dispatch(setIntervals(intervals))
      })
  }
}