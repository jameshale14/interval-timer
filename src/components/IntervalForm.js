import React from 'react'
import PropTypes from 'prop-types'

export default class IntervalForm extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      stepError: undefined,
      saveError: undefined,
      name: props.interval ? props.interval.name : '',
      steps: props.interval ? props.interval.steps : [],
      newStep: {
        type: 'Activity',
        name: '',
        duration: ''
      }
    }
  }

  handleAddStep = (e) => {
    e.preventDefault()
    if (this.state.newStep.type == '' || this.state.newStep.name == '' || this.state.newStep.duration == 0) {
      this.setState(() => ({ stepError: 'Please fill in all fields' }))
    } else {
      this.setState((prevState) => {
        const newStep = { ...this.state.newStep }
        return {
          steps: [...prevState.steps, newStep],
          stepError: undefined
        }
      })
      // this.setState(() => ({  }))
    }
  }

  onTypeChange = (e) => {
    const type = e.target.value
    this.setState((prevState) => {
      const newStep = prevState.newStep
      newStep.type = type
      return { newStep }
    })
  }

  onNameChange = (e) => {
    const name = e.target.value
    this.setState((prevState) => {
      const newStep = prevState.newStep
      newStep.name = name
      return { newStep }
    })
  }

  onIntervalNameChange = (e) => {
    const name = e.target.value
    this.setState(() => ({ name }))
  }

  onDurationChange = (e) => {
    const duration = e.target.value
    this.setState((prevState) => {
      const newStep = prevState.newStep
      newStep.duration = duration
      return { newStep }
    })
  }

  handleOnSave = (e) => {
    e.preventDefault()

    if (!this.state.name || this.state.steps.length === 0) {
      this.setState(() => ({ saveError: 'Please provide a routine name and interval steps.' }))
    } else {
      this.setState(() => ({ saveError: undefined }))

      const interval = {
        name: this.state.name,
        steps: this.state.steps
      }

      this.props.onSubmit(interval)
    }

  }

  handleRemoveStep = (e) => {
    e.preventDefault()
    const stepIndex = parseInt(e.currentTarget.value)
    this.setState((prevState) => {
      const steps = prevState.steps.filter((step, index) => {
        return index != stepIndex
      })

      return ({
        steps
      })

    })
  }

  render() {
    return (
      <div>
        <button id='saveInterval' onClick={this.handleOnSave}>Save</button>
        {this.state.saveError && <p>{this.state.saveError}</p>}
        <input
          type='text'
          placeholder='name'
          value={this.state.name}
          onChange={this.onIntervalNameChange}
        />
        {
          this.state.steps.map((step, stepIndex) => {
            return (
              <div key={stepIndex} >
                <p>Type: {step.type}</p>
                <p>Name: {step.name}</p>
                <p>Duration: {step.duration}</p>
                <button onClick={this.handleRemoveStep} value={stepIndex}>Remove</button>
              </div>
            )
          })
        }
        <div>
          {this.state.stepError && <p>{this.state.stepError}</p>}
          <form onSubmit={this.handleAddStep}>
            <select
              value={this.state.newStep.type}
              onChange={this.onTypeChange}
              onBlur={this.onTypeChange}
            >
              <option value='Activity'>Activity</option>
              <option value='Rest'>Rest</option>
            </select>

            <input
              type='text'
              placeholder='Step name'
              value={this.state.newStep.name}
              onChange={this.onNameChange}
            />
            <input
              type='number'
              placeholder='Duration (seconds)'
              step='1'
              value={this.state.newStep.duration}
              onChange={this.onDurationChange}
            />
            <button onClick={this.handleAddStep}>Add Step</button>
          </form>
        </div>
      </div>
    )
  }
}

IntervalForm.propTypes = {
  interval: PropTypes.object,
  onSubmit: PropTypes.func
}
