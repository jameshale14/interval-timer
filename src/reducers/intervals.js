export default (state = {}, action) => {
  switch (action.type) {
    case 'GET_INTERVALS':
      return state
    case 'CREATE_INTERVAL':
      return [...state, action.interval]
    case 'UPDATE_INTERVAL':
      return state.map((interval) => {
        if (interval.id === action.id) {
          return {
            ...interval,
            ...action.updates
          }
        } else {
          return interval
        }
      })
    case 'DELETE_INTERVAL':
      return state.filter((interval) => {
        return interval.id !== action.id
      })
    case 'SET_INTERVALS':
      return action.intervals
    default:
      return state
  }
}