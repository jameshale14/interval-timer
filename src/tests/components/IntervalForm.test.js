import React from 'react'
import { shallow } from 'enzyme'
import IntervalForm from '../../components/IntervalForm'
import { intervals } from '../fixtures/intervals'

let wrapper, onSubmit

beforeEach(() => {
  onSubmit = jest.fn()
  wrapper = shallow(<IntervalForm onSubmit={onSubmit} />)
})

test('should render empty IntervalForm correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should render pre-filled IntervalForm when sending interval data', () => {
  const wrapper = shallow(<IntervalForm onSubmit={onSubmit} interval={intervals[1]} />)
  expect(wrapper).toMatchSnapshot()
})

test('should add step to step list and render to screen when added from StepForm', () => {
  const newStep = {
    type: 'Activity',
    name: 'Squats',
    duration: 30
  }

  wrapper.find('StepForm').prop('onSubmit')(newStep)
  expect(wrapper.state('steps')).toEqual([newStep])
  expect(wrapper).toMatchSnapshot()
})

test('should fire a save event when saved with valid submission', () => {
  wrapper.find('input').at(0).simulate('change', {
    target: { value: 'New Interval Routine' }
  })

  const newStep = {
    type: 'Activity',
    name: 'Push-ups',
    duration: 30
  }

  wrapper.find('StepForm').prop('onSubmit')(newStep)

  wrapper.find('button').at(0).simulate('click', {
    preventDefault: () => { }
  })
  expect(onSubmit).toHaveBeenLastCalledWith({
    name: 'New Interval Routine',
    steps: [
      {
        type: 'Activity',
        name: 'Push-ups',
        duration: 30
      }
    ]
  })
})


test('should show an error when saving without enough data', () => {
  wrapper.find('button').at(0).simulate('click', {
    preventDefault: () => { }
  })
  expect(wrapper.state().saveError.length).toBeGreaterThan(0)
  expect(wrapper).toMatchSnapshot()
})

test('should remove step when pressed', () => {
  const wrapperWithDetail = shallow(<IntervalForm onSubmit={onSubmit} interval={intervals[0]} />)

  wrapperWithDetail.find('button').at(2).simulate('click', {
    preventDefault: () => { },
    currentTarget: { value: 1 }
  })

  expect(wrapperWithDetail.state().steps).toEqual([intervals[0].steps[0], intervals[0].steps[2], intervals[0].steps[3]])
  expect(wrapperWithDetail).toMatchSnapshot()

})