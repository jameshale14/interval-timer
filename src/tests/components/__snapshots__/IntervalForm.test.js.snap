// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add step to step list and render to screen when added from StepForm 1`] = `
<div>
  <button
    id="saveInterval"
    onClick={[Function]}
  >
    Save
  </button>
  <input
    onChange={[Function]}
    placeholder="name"
    type="text"
    value=""
  />
  <div
    key="0"
  >
    <p>
      Type: 
      Activity
    </p>
    <p>
      Name: 
      Squats
    </p>
    <p>
      Duration: 
      30
    </p>
    <button
      onClick={[Function]}
      value={0}
    >
      Remove
    </button>
  </div>
  <StepForm
    onSubmit={[Function]}
  />
</div>
`;

exports[`should remove step when pressed 1`] = `
<div>
  <button
    id="saveInterval"
    onClick={[Function]}
  >
    Save
  </button>
  <input
    onChange={[Function]}
    placeholder="name"
    type="text"
    value="fitness workout 1"
  />
  <div
    key="0"
  >
    <p>
      Type: 
      activity
    </p>
    <p>
      Name: 
      press-ups
    </p>
    <p>
      Duration: 
      20000
    </p>
    <button
      onClick={[Function]}
      value={0}
    >
      Remove
    </button>
  </div>
  <div
    key="1"
  >
    <p>
      Type: 
      activity
    </p>
    <p>
      Name: 
      crunches
    </p>
    <p>
      Duration: 
      20000
    </p>
    <button
      onClick={[Function]}
      value={1}
    >
      Remove
    </button>
  </div>
  <div
    key="2"
  >
    <p>
      Type: 
      rest
    </p>
    <p>
      Name: 
      rest
    </p>
    <p>
      Duration: 
      10000
    </p>
    <button
      onClick={[Function]}
      value={2}
    >
      Remove
    </button>
  </div>
  <StepForm
    onSubmit={[Function]}
  />
</div>
`;

exports[`should render empty IntervalForm correctly 1`] = `
<div>
  <button
    id="saveInterval"
    onClick={[Function]}
  >
    Save
  </button>
  <input
    onChange={[Function]}
    placeholder="name"
    type="text"
    value=""
  />
  <StepForm
    onSubmit={[Function]}
  />
</div>
`;

exports[`should render pre-filled IntervalForm when sending interval data 1`] = `
<div>
  <button
    id="saveInterval"
    onClick={[Function]}
  >
    Save
  </button>
  <input
    onChange={[Function]}
    placeholder="name"
    type="text"
    value="meditation focus"
  />
  <div
    key="0"
  >
    <p>
      Type: 
      activity
    </p>
    <p>
      Name: 
      focus
    </p>
    <p>
      Duration: 
      30000
    </p>
    <button
      onClick={[Function]}
      value={0}
    >
      Remove
    </button>
  </div>
  <div
    key="1"
  >
    <p>
      Type: 
      rest
    </p>
    <p>
      Name: 
      rest
    </p>
    <p>
      Duration: 
      15000
    </p>
    <button
      onClick={[Function]}
      value={1}
    >
      Remove
    </button>
  </div>
  <div
    key="2"
  >
    <p>
      Type: 
      activity
    </p>
    <p>
      Name: 
      focus
    </p>
    <p>
      Duration: 
      30000
    </p>
    <button
      onClick={[Function]}
      value={2}
    >
      Remove
    </button>
  </div>
  <div
    key="3"
  >
    <p>
      Type: 
      rest
    </p>
    <p>
      Name: 
      rest
    </p>
    <p>
      Duration: 
      15000
    </p>
    <button
      onClick={[Function]}
      value={3}
    >
      Remove
    </button>
  </div>
  <StepForm
    onSubmit={[Function]}
  />
</div>
`;

exports[`should show an error when saving without enough data 1`] = `
<div>
  <button
    id="saveInterval"
    onClick={[Function]}
  >
    Save
  </button>
  <p>
    Please provide a routine name and interval steps.
  </p>
  <input
    onChange={[Function]}
    placeholder="name"
    type="text"
    value=""
  />
  <StepForm
    onSubmit={[Function]}
  />
</div>
`;
